theano/sandbox/gpuarray/opt.py

```diff
@@ -146,35 +166,81 @@ def apply(self, fgraph):
 
             if (len(input.clients) == 1 and
                 (input.clients[0][0] == 'output' or
-                 input.clients[0][0].op == gpu_from_host)):
+                 isinstance(input.clients[0][0].op, GpuFromHost))):
                 continue
 
+            ctx_name = getattr(input.tag, 'context_name', None)
             try:
-                new_input = host_from_gpu(gpu_from_host(input))
+                new_input = host_from_gpu(GpuFromHost(ctx_name)(input))
                 fgraph.replace_validate(input, new_input,
                                         "InputToGpuOptimizer")
             except TypeError:
                 # This could fail if the inputs are not TensorTypes
                 pass
+            except ValueError:
```

Comment:
I'm pretty sure I did that, actually.  I may have forgot some places where it should be changed though, like here.  But I made the new exception inherit from ValueError to avoid problems in the meantime.